cntk: fix build and add comments

glennpj opened the pull request.

glennpj committed:
cntk: fix and deprecate

alalazo reviewed (dismissed):
LGTM, but wanted to double check that the intent is to deprecate _all versions_

glennpj commented:
> LGTM, but wanted to double check that the intent is to deprecate _all versions_

Perhaps the comments that I put in the file are sufficient rather than deprecating?

alalazo commented:
> Perhaps the comments that I put in the file are sufficient rather than deprecating?

I am asking since currently one needs to install with `--deprecated` to have Spack use a deprecated version without being prompted at fetching time:
```console
$ spack install openssl@1.1.1l
==> Warning: using "openssl@1.1.1l" which is a deprecated version
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/berkeley-db-18.1.40-jjcdgicmjttpteerizjgcko7ivhxp7yh
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/libiconv-1.16-klh4setxxl7qui3sr56bjraw3qskydqu
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/pkgconf-1.8.0-k7oj7j7cv6qkren5nu2zpaknr3at4d2z
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/zlib-1.2.11-h7tvzacqculnkfs44qyfaizurnqsfvdr
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/diffutils-3.8-b6dr6dxubin6lktaamjxqwdvwii7ngrt
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/ncurses-6.2-vc37m24s7t6af7vwhhlvhd22lmbchqib
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/bzip2-1.0.8-iwjq7ca2gxdipsqojnb2h5xn3lxz2orb
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/readline-8.1-autymkx77iesq7hbpwiwg5il2j52zrer
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/gdbm-1.19-spq3vzbvhjcnu7admmbrijoami454pgl
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/perl-5.34.0-esrjoliefurio2rnhkuzivvhyawkcdzq
==> Installing openssl-1.1.1l-roiqch4bpn5kopz6zcnde5awjb6r62c3
==> No binary for openssl-1.1.1l-roiqch4bpn5kopz6zcnde5awjb6r62c3 found: installing from source
==> Warning: openssl@1.1.1l is deprecated and may be removed in a future Spack release.
==>   Fetch anyway? [y/N] 
```
Whether or not that is what we want I think it's a decision of the maintainers of the package and the community. I never used `cntk` so I have no opinion here. Maybe @adamjstewart has some thoughts?

adamjstewart commented on var/spack/repos/builtin/packages/cntk/package.py:
```diff
@@ -49,6 +54,26 @@ class Cntk(Package):
+
+    def patch(self):
+        if 'protobuf+shared' in self.spec:
+            library_suffix = 'so'
```
Is this also true on macOS?

glennpj replied:
This will not be necessary at all with other changes being made.

adamjstewart commented on var/spack/repos/builtin/packages/cntk/package.py:
```diff
@@ -49,6 +54,26 @@ class Cntk(Package):
+        protobuf_ld_flags = self.spec['protobuf'].libs.ld_flags
+
+        filter_file(r'(protobuf_check=)lib/(libprotobuf\.)a',
+                    r'\1{0}/\2{1}'.format(protobuf_libdir, library_suffix),
```
You should be able to use `spec['protobuf'].libs[0]` if you just want the full path to the library file

glennpj replied:
Due to the way the check is in the configure script, it needs to be split. But your comment gave me an idea of how to do this better and eliminate the need for the suffix variable.

adamjstewart reviewed (commented):
Many of the packages in Spack are no longer actively developed. We don't necessarily need to deprecate and remove this package, there might be people who still use it for reproducibility of old ML experiments.

glennpj commented:
> Many of the packages in Spack are no longer actively developed. We don't necessarily need to deprecate and remove this package, there might be people who still use it for reproducibility of old ML experiments.

Yep, I went to the spack docs to reread the deprecation section. I will back that out.

glennpj committed:
Fix some issues with PR

- remove deprecation flags
- remove library_suffix variable
- split path to protobuf library into components
- redo filter_file accordingly

glennpj committed:
fix indentation
